dsl/simulation/control.ml

```diff
@@ -0,0 +1,165 @@
+    | Strict plst, h :: t ->
+        let pt, s', time =
+          z_in_aux (List.nth plst h) t (Cons (h, addr)) s pkt
+        in
+        ((h, Rank.create_for_pkt (float_of_int h) pkt) :: pt, s', time)
+    | RoundRobin plst, h :: t ->
+        let n = List.length plst in
+        let r_i = sprintf "%s_r_%d" prefix h in
+        let rank = State.lookup r_i s in
+        let s' = State.rebind r_i (rank +. float_of_int n) s in
+        let pt, s'', time =
+          z_in_aux (List.nth plst h) t (Cons (h, addr)) s' pkt
+        in
+        ((h, Rank.create_for_pkt rank pkt) :: pt, s'', time)
+    | WeightedFair plst, h :: t ->
+        let lf, w = (sprintf "%s_lf_%d" prefix h, sprintf "%s_w_%d" prefix h) in
+        let weight = State.lookup w s in
+        let rank =
+          match State.lookup_opt lf s with
+          | Some v -> max (pkt |> Packet.time |> Time.to_float) v
+          | None -> pkt |> Packet.time |> Time.to_float
+        in
+        let s' = State.rebind lf (rank +. (Packet.len pkt /. weight)) s in
+        let pt, s'', time =
+          z_in_aux (List.nth plst h |> fst) t (Cons (h, addr)) s' pkt
+        in
+        ((h, Rank.create_for_pkt rank pkt) :: pt, s'', time)
+    | _ -> failwith "ERROR: unreachable branch"
+  in
+  match route_pkt_opt p pkt with
+  | Some rank_less_path -> z_in_aux p rank_less_path Eps s pkt
+  | None -> raise (RoutingError pkt)
+
+let z_out p s pkt =
+  let rec z_out_aux (p : Frontend.Policy.t) rank_less_path addr s pkt =
+    let prefix = addr_to_string addr in
+
+    match (p, rank_less_path) with
+    | Class _, [] -> s
+    | Fifo plst, h :: t | Strict plst, h :: t ->
+        z_out_aux (List.nth plst h) t (Cons (h, addr)) s pkt
+    | WeightedFair plst, h :: t ->
+        z_out_aux (List.nth plst h |> fst) t (Cons (h, addr)) s pkt
+    | RoundRobin plst, h :: t ->
+        let n = List.length plst in
+        let who_skip pop turn =
+          let rec who_skip_aux t acc =
+            if t = pop then acc else who_skip_aux ((t + 1) mod n) (t :: acc)
+          in
+          who_skip_aux turn []
+        in
+        let turn = State.lookup (sprintf "%s_turn" prefix) s in
+        let s' =
+          State.rebind (sprintf "%s_turn" prefix)
+            ((h + 1) mod n |> float_of_int)
+            s
+        in
+        let skipped = who_skip h (int_of_float turn) in
+        let f s i =
+          let r_i = sprintf "%s_r_%d" prefix i in
+          State.rebind r_i (State.lookup r_i s +. float_of_int n) s
+        in
+        let s'' = List.fold_left f s' skipped in
+        z_out_aux (List.nth plst h) t (Cons (h, addr)) s'' pkt
+    | _ -> failwith "ERROR: unreachable branch"
+  in
+  match route_pkt_opt p pkt with
+  | Some rank_less_path -> z_out_aux p rank_less_path Eps s pkt
+  | None -> raise (RoutingError pkt)
+
+let of_policy p =
+  {
+    q = p |> Topo.of_policy |> Pieotree.create;
+    s = init_state p;
+    z_in = z_in p;
+    z_out = z_out p;
```

Comment:
Reading these lines is making me realize that maybe the _methods_ `z_in` and `z_out` above should be renamed into `z_in_of_policy` and `z_in_of_policy`.

Reply:
Done by [8614930](https://github.com/cucapra/packet-scheduling/pull/57/commits/8614930daa6d72a532e752fb6760bb1b7b0e320a)